contracts/modules/capital/YieldDeposit.sol

```diff
@@ -0,0 +1,182 @@
+    }
+    if (amount <= 0) {
+      revert InvalidDepositAmount();
+    }
+
+    token.safeTransferFrom(msg.sender, address(this), amount);
+    deposits[msg.sender] += amount;
+    initialRates[msg.sender] = getCurrentTokenPrice();
+
+    totalPrincipal += amount;
+    uint coverAmount = updateCoverAmount(msg.sender);
+
+    emit TokenDeposited(msg.sender, amount, coverAmount);
+  }
+
+  /**
+   * @notice Withdraws the entire principal amount previously deposited by the caller.
+   * @dev Transfers the deposited tokens back to the caller and resets their deposited balance to zero.
+   *      This action is only possible if the caller has a positive deposited balance.
+   */
+  function withdraw() external nonReentrant {
+
+    uint principalAmount = deposits[msg.sender];
+    if (principalAmount <= 0) {
+      revert InsufficientDepositForWithdrawal();
+    }
+
+    uint currentRate = getCurrentTokenPrice();
+    uint withdrawAmount = (principalAmount * initialRates[msg.sender]) / currentRate;
+
+    totalPrincipal -= principalAmount;
+    deposits[msg.sender] = 0;
+    coverAmounts[msg.sender] = 0;
+    initialRates[msg.sender] = 0;
+
+    token.safeTransfer(msg.sender, withdrawAmount);
+
+    emit TokenWithdrawn(msg.sender, withdrawAmount);
+  }
+
+  /**
+   * @notice Allows the contract owner to withdraw the accumulated yield.
+   * @dev The yield is defined as the difference between the current token balance of the contract
+   *      and the total principal. Reverts with `NoYieldAvailable` if there is no yield.
+   */
+  function withdrawAvailableYield() external onlyOwner nonReentrant {
+
+    uint availableYield = getAvailableYield();
+    if (availableYield == 0) {
+      revert NoYieldAvailable();
+    }
+
+    token.safeTransfer(owner(), availableYield);
+  }
+
+  /**
+   * @notice Retrieves the current yield of the contract based on the token's price and total deposits.
+   * @dev Calculates yield as the difference between current token value and total principal if positive.
+   * @return The current yield as an unsigned integer. Returns zero if the current value is less than the total principal.
+   */
+  function getAvailableYield() public view returns (uint) {
+
+    uint currentContractValue = getCurrentTokenValue();
+    if (currentContractValue < totalPrincipal) {
+      return 0;
+    }
+    return currentContractValue - totalPrincipal;
+  }
+
+  /**
+   * @notice Calculates the current market value of the tokens held by the contract.
+   * @dev Fetches the latest token price from the Chainlink price feed and multiplies it by the number of tokens in the contract.
+   * @return The total value of the tokens held by the contract as an unsigned integer.
+   */
+  function getCurrentTokenValue() private view returns (uint) {
+
+    uint totalTokens = token.balanceOf(address(this));
+    uint currentTokenPrice = getCurrentTokenPrice();
+
+    return (totalTokens * currentTokenPrice) / (10 ** tokenDecimals);
+  }
```

Comment:
User withdrawal

- pull the latest oracle value
- withdrawable amount = deposit * initial_rate / current_rate
  - ex: 100 -> 100 * 1.2 / 1.3 = 92.30769230769231